db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/FromBaseTable.java

```diff
@@ -561,6 +566,48 @@ public boolean isCoveringIndex(ConglomerateDescriptor cd) throws StandardExcepti
         return coveringIndex;
     }
 
+    private boolean areAllReferencingExprsCoveredByIndex(IndexDescriptor id) throws StandardException {
+        if (referencingExpressions == null || referencingExpressions.isEmpty()) {
+            return false;
+        }
+        return getRefExprIndexPositions(id).size() == referencingExpressions.size();
```

Comment:
This check is too simple to cover all combinations.  For example:

> create table t1 (a int, b int, primary key(a));
> create index t1_idx on t1(a+1, b);
> explain select count(*) from t1 --splice-properties index=t1_idx
> where a+1 > 1 and b+b > 1;
>       ->  GroupBy(n=4,totalCost=38.017,outputRows=1,outputHeapSize=0 B,partitions=1)
>         ->  ProjectRestrict(n=3,totalCost=30.413,outputRows=4,outputHeapSize=7 B,partitions=1)
>           ->  ProjectRestrict(n=2,totalCost=30.413,outputRows=4,outputHeapSize=7 B,partitions=1,preds=[((B[1:2] + B[1:2]) > 1)])
>             ->  IndexLookup(n=1,totalCost=30.413,outputRows=7,outputHeapSize=7 B,partitions=1)
>               ->  IndexScan[T1_IDX(7249)](n=0,totalCost=4.013,scannedRows=7,outputRows=7,outputHeapSize=7 B,partitions=1,baseTable=T1(7232),preds=[((A[1:1] + 1) > 1)])
> 

Even though column B is in the index, an expensive IndexLookup operation is done in order to evaluate b+b > 1.  Without expression-based indexes, the reference to b+b does not cause an IndexLookup.

I'm not sure how we should handle this, perhaps by not marking rc.isReferenced for column A in this case, or maybe detecting whether all column references of a given column only occur as matched index expressions.  But having to do IndexLookup is very expensive so we want to always avoid it if possible.

Reply:
I agree that the check is simple at the moment. However, doing something similar on expressions is probably more difficult than on column references. In general, we can avoid IndexLookup in the following case:

> create index t1_idx on t1(a+1);
> select * from t1 where 2*a+2 > 3;  -- possible to utilize (a+1) column without IndexLookup

I think I can take care of the case you mentioned in the example, in which a column reference is an index expression. But for more general cases, I haven't got clear ideas on that.

Reply:
Can you open a Jira to take care of the general case?  I believe regular indexes can handle such expressions without doing an IndexLookup.

Reply:
Yes, I'll open a Jira ticket to track this issue.

Reply:
Here it is: https://splicemachine.atlassian.net/browse/DB-10653